src/caffe/layers/reshape_layer.cpp

```diff
@@ -0,0 +1,20 @@
+#include <vector>
+
+#include "caffe/common_layers.hpp"
+#include "caffe/layer.hpp"
+
+namespace caffe {
+
+template <typename Dtype>
+void ReshapeLayer<Dtype>::Reshape(const vector<Blob<Dtype>*>& bottom,
+      const vector<Blob<Dtype>*>& top) {
+  top[0]->Reshape(this->layer_param_.reshape_param().shape());
```

Comment:
Right, to me it feels like this should go in `LayerSetUp`. If special options were added so that top shape depended on bottom shape, then `Reshape` would be needed. But that isn't the case in this code.